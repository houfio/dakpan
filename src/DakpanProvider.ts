import { createElement, ReactElement } from 'react';

import { DakpanMultiProviderProps } from './types';

export const DakpanProvider = ({ provide, children }: DakpanMultiProviderProps) => (
  provide.reduce((previous, current) => createElement(current, { children: previous }), children as ReactElement<any>)
);
