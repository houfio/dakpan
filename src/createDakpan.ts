import {
  Component,
  ComponentClass,
  ConsumerProps,
  createContext,
  ProviderProps,
  ReactElement,
  ReactNodeArray,
  ReactPortal,
  StatelessComponent
} from 'react';

import { createConsumer } from './createConsumer';
import { createProvider } from './createProvider';
import { Actions, DakpanProviderProps, GetState, MappedActions, SetState } from './types';
import { withDakpan } from './withDakpan';

export const createDakpan = <S>(initialState: S) => <A extends Actions<S>>(actions: A) => {
  let getState: GetState<S> | undefined;
  let setState: SetState<S, DakpanProviderProps> | undefined;

  const mergeAction = (result: Partial<S>) => {
    if (!getState || !setState) {
      throw new Error('Provider unmounted before action could complete');
    } else if (typeof result !== 'object') {
      throw new Error('Actions may only return objects');
    }

    const newState = {
      ...getState() as any,
      ...result as any
    };

    setState(newState);

    return newState;
  };

  const { Provider, Consumer } = createContext<S>(initialState);
  const mappedActions = Object.keys(actions).reduce(
    (result, key) => {
      const e = (...args: any[]) => () => {
        const action = actions[key];

        if (!getState || !setState) {
          throw new Error('You may not dispatch an action before its provider is mounted');
        }

        const state = getState();
        const result = action(...args)(state);

        if (result instanceof Promise) {
          return result.then(mergeAction);
        }

        return mergeAction(result);
      };

      const action: any = (...args: any[]) => e(...args)();
      action.e = e;

      return {
        ...result,
        [key]: action
      };
    },
    {}
  ) as MappedActions<S, A>;

  const provider = createProvider(Provider, initialState, (get, set) => {
    getState = get;
    setState = set;
  });
  const consumer = createConsumer(Consumer, mappedActions);

  return {
    Provider: provider,
    Consumer: consumer,
    actions: mappedActions,
    withDakpan: withDakpan(consumer)
  };
};
