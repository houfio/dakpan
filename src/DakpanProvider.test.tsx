import { mount } from 'enzyme';
import * as React from 'react';

import { createDakpan } from './createDakpan';
import { DakpanProvider } from './DakpanProvider';

it('provides multiple dakpan states', () => {
  const { Provider: FirstProvider } = createDakpan({
    hello: 'world'
  })({
    append: (append: string) => ({ hello }) => ({
      hello: `${hello}${append}`
    })
  });

  const { Provider: SecondProvider, Consumer: SecondConsumer, actions: secondActions } = createDakpan({
    test: 0
  })({
    increment: () => ({ test }) => ({
      test: test + 1
    })
  });

  const wrapper = mount(
    <DakpanProvider provide={[FirstProvider, SecondProvider]}>
      <SecondConsumer>
        {({ test }) => (
          <span>{test}</span>
        )}
      </SecondConsumer>
    </DakpanProvider>
  );

  expect(secondActions.increment).not.toThrow();
  expect(wrapper.update()).toMatchSnapshot();
});
