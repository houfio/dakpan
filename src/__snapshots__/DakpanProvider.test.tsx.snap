// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should provide multiple dakpan states 1`] = `
<Component
  provide={
    Array [
      [Function],
      [Function],
    ]
  }
>
  <Provider>
    <Provider>
      <Consumer>
        <span>
          1
        </span>
      </Consumer>
    </Provider>
  </Provider>
</Component>
`;
